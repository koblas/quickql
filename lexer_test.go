package quickql

import (
	"strings"
	"testing"

	"github.com/alecthomas/participle/v2/lexer"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"
)

func TestBasicLexer(t *testing.T) {
	cases := []struct {
		input string
	}{
		{input: "hello=world"},
		{input: "hello = world"},
		{input: "hello:world"},
		{input: "hello<world"},
		{input: "hello = [world]"},
		{input: "hello != [world]"},
		{input: "hello:66e839032fb119d31dc9c968"},
		{input: `foo bar`},
		{input: `"foo bar"`},
		{input: `"foo \"bar\""`},
		{input: `foo OR bar`},
		{input: "-10"},
		{input: "-10.7"},
		{input: "10.7"},
		{input: "hello:world\nfoo:bar"},
		{input: "-hello:world"},
		{input: `name = "Bob" and (age > 20 or age < 5)`},
	}

	queryLexer := NewLexer()

	symbols := queryLexer.Symbols()
	names := map[lexer.TokenType]string{}
	for k, v := range symbols {
		names[v] = k
	}

	for _, c := range cases {
		t.Run(c.input, func(t *testing.T) {
			l, err := queryLexer.Lex("", strings.NewReader(c.input))
			require.NoError(t, err)

			tokens, err := lexer.ConsumeAll(l)
			require.NoError(t, err)

			snaps.MatchSnapshot(t, DebugOutput(tokens))
		})
	}
}

func TestFailLexer(t *testing.T) {
	cases := []struct {
		input string
	}{
		{input: `"foo`},
		{input: `'test\`},
		{input: `foo!bar`},
		{input: `foo==bar`},
	}

	queryLexer := NewLexer()

	symbols := queryLexer.Symbols()
	names := map[lexer.TokenType]string{}
	for k, v := range symbols {
		names[v] = k
	}

	for _, c := range cases {
		t.Run(c.input, func(t *testing.T) {
			l, err := queryLexer.Lex("", strings.NewReader(c.input))
			require.NoError(t, err)

			_, err = lexer.ConsumeAll(l)
			require.Error(t, err)
		})
	}
}

func TestSimple(t *testing.T) {
	input := `(fish!=food)`
	// input := `-hello:world`
	l, err := queryLexer.Lex("", strings.NewReader(input))
	require.NoError(t, err)

	tokens, err := lexer.ConsumeAll(l)
	require.NoError(t, err)
	require.NotEmpty(t, tokens)
}
