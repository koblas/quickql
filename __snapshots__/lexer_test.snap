
[TestBasicLexer/hello=world - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"=\"}", "Token{IDENT, \"world\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/hello_=_world - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"=\"}", "Token{IDENT, \"world\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/hello:world - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"=\"}", "Token{IDENT, \"world\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/hello<world - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"<\"}", "Token{IDENT, \"world\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/hello_=_[world] - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"=\"}", "Token{VALUE, \"[world]\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/hello_!=_[world] - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"!=\"}", "Token{VALUE, \"[world]\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/hello:66e839032fb119d31dc9c968 - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"=\"}", "Token{VALUE, \"66e839032fb119d31dc9c968\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/foo_bar - 1]
[]string{"Token{IDENT, \"foo\"}", "Token{IDENT, \"bar\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/"foo_bar" - 1]
[]string{"Token{STRING, \"foo bar\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/"foo_\"bar\"" - 1]
[]string{"Token{STRING, \"foo \\\"bar\\\"\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/foo_OR_bar - 1]
[]string{"Token{IDENT, \"foo\"}", "Token{IDENT, \"OR\"}", "Token{IDENT, \"bar\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/-10 - 1]
[]string{"Token{VALUE, \"-10\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/-10.7 - 1]
[]string{"Token{VALUE, \"-10.7\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/10.7 - 1]
[]string{"Token{VALUE, \"10.7\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/hello:world_foo:bar - 1]
[]string{"Token{IDENT, \"hello\"}", "Token{VALUE, \"=\"}", "Token{IDENT, \"world\"}", "Token{IDENT, \"foo\"}", "Token{VALUE, \"=\"}", "Token{IDENT, \"bar\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/-hello:world - 1]
[]string{"Token{VALUE, \"NOT\"}", "Token{IDENT, \"hello\"}", "Token{VALUE, \"=\"}", "Token{IDENT, \"world\"}", "Token{EOF, \"\"}"}
---

[TestBasicLexer/name_=_"Bob"_and_(age_>_20_or_age_<_5) - 1]
[]string{"Token{IDENT, \"name\"}", "Token{VALUE, \"=\"}", "Token{STRING, \"Bob\"}", "Token{IDENT, \"and\"}", "Token{LPAREN, \"(\"}", "Token{IDENT, \"age\"}", "Token{VALUE, \">\"}", "Token{VALUE, \"20\"}", "Token{IDENT, \"or\"}", "Token{IDENT, \"age\"}", "Token{VALUE, \"<\"}", "Token{VALUE, \"5\"}", "Token{RPAREN, \")\"}", "Token{EOF, \"\"}"}
---
